[pypi] Fix cleanup logic for multiple branches

ZuleykhaPavlichenkova-TomTom opened the pull request:
Python Twine upload [keeps failing](https://github.com/duckdb/duckdb/actions/runs/13824340583/job/38685443491) nightly:
```
ERROR    HTTPError: 400 Bad Request from https://upload.pypi.org/legacy/        
         Project size too large. Limit for project 'duckdb' total size is 30 GB.
         See https://pypi.org/help/#project-size-limit    
```

The script we use to cleanup the space on pypi was designed for single branch, but now we upload from two branches. We need to delete old dev builds from both of them, but really do that only for dev builds created from `main` branch.

@carlopi suggested a cleanup logic to fix it and helped me to start up with implementation.

The fix is to update the cleanup process to run separately for each branch.
The logic is to shift from cleaning files older than X days on the `main` branch to cleaning files older than X/2 days on both the `main` and `v1.2-` branches.
This PR changes number of days to retain dev releases to 5 (was 10, but now it's 5 per both main and v1.2-), adds `DUCKDB_BASE_VERSION` env variable to filter out the pkg versions which we don't want to keep on pypi.

To test it run: `PYPI_CLEANUP_USERNAME=user PYPI_CLEANUP_PASSWORD=psw PYPI_CLEANUP_OTP=otp DUCKDB_BASE_VERSION=<major.minor> python3 scripts/pypi_cleanup.py` replacing `<major.minor>` with duckdb's base version (actual versions for now are 1.2 and 1.3 for `main`).

Should exit if `DUCKDB_BASE_VERSION` is empty, shorter than 3 symbols, contains more or less than 1 dots.
When the version is set to one of actual versions, expected that the value of `pkg_vers` variable contains the list of the dev releases names to delete.

ZuleykhaPavlichenkova-TomTom committed:
set DUCKDB_BASE_VERSION env variable to make pypi_cleanup.py script clean up builds created by both (release and main) branches

ZuleykhaPavlichenkova-TomTom committed:
set retain days to 5 so there could be 5 builds per release and main;
add check that duckdb version is set and has correct length;
filter pkg_vers using a pattern containing a duckdb_version_base

ZuleykhaPavlichenkova-TomTom committed:
check that duckdb version contains only 1 dot

hmeriann commented:
@Mytherin could you please delete some of the latest `PRE-RELEASE` versions from [pypi](https://pypi.org/project/duckdb/#history)? Otherwise it fails again, because there is no space for new uploads.

carlopi commented on .github/workflows/Python.yml:
```diff
@@ -432,14 +432,15 @@ jobs:
        PYPI_CLEANUP_USERNAME: 'mytherin'
        PYPI_CLEANUP_PASSWORD: ${{secrets.PYPI_CLEANUP_PASSWORD}}
        PYPI_CLEANUP_OTP: ${{secrets.PYPI_CLEANUP_OTP}}
+       DUCKDB_BASE_VERSION: ${{github.ref == 'refs/heads/main' && '1.3' || '1.2'}}
```
This part is the weirder one, but I am not sure if / what is a more proper alternative. This would need to be revised / changed moving forward. I don't have a proper answer, possibly it's fine and to be generalised later (together with removing the hardcoded branch name), first checking this work as intended.

carlopi reviewed (approved):
Thanks

I think this is OK for now, and worth getting sorted out.

We might need to generalise somehow, probably a variable, a JSON string holding the mapping between tested branches AND targeted versions.
That would be a cool +1 to clean this up, but maybe first worth testing.

carlopi commented:
Actually, could you make this PR (only the 3 relevant commits) vs branch `v1.2-histrionicus` ?
That way we avoid duplication / workflow diverging.

Mytherin commented:
Thanks for the PR!

This looks a bit brittle/error prone to me. I would suggest simplifying this and - instead of having the script behave differently depending on where it is run, we can make the script itself a bit more clever in which releases are deleted.

I would propose to modify the script to keep the last 3 dev releases of the next major version (`1.3.X`) and the last 3 dev releases of the next minor version (`1.2.2-X`).

This moves the behavior from relying on "retention days" - which should also prevent us from deleting all releases when builds fail for a few days in a row, and further reduce brittleness here.

ZuleykhaPavlichenkova-TomTom committed:
replase retention days with the number of version we'd like to keep
when there is a release, remove all its pre-release versions

ZuleykhaPavlichenkova-TomTom committed:
remove DUCKDB_BASE_VERSION env variable since we don't pass it to the cleanup script anymore

hmeriann commented:
@Mytherin now the retention days are replaced with the number of versions to keep (3), so we don't need to pass the `DUCKDB_BASE_VERSION` variable to the script. Also, when a new version've been released, the script should remove all of its PRE_RELEASE versions.

Mytherin commented on scripts/pypi_cleanup.py:
```diff
@@ -13,9 +14,9 @@
-patterns = [re.compile(r".*\.dev\d+$")]
+# how many dev versions to keep - 3 for each not yet released version
+how_many_dev_versions_to_keep = 3
+patterns = [re.compile(rf"^{duckdb_version_base}.*\.dev\d+$")]
```
`{duckdb_version_base}` shouldn't be here anymore right?

Mytherin commented on scripts/pypi_cleanup.py:
```diff
@@ -13,9 +14,9 @@
 
 # deletes old dev wheels from pypi. evil hack.
-# how many days to retain dev releases - all dev releases older than 10 days are deleted
-retain_days = 10
```
This variable still appears to be used in the script so this doesn't work right now

Mytherin commented on scripts/pypi_cleanup.py:
```diff
@@ -118,12 +119,18 @@ def run(self):
+            for version_key, versions in version_dict.items():
+                if version_key in releases_by_date.keys():
+                    pkg_vers.extend(versions)
+                pkg_vers.extend(versions[:-how_many_dev_versions_to_keep])
```
Should this be `else:`?

Mytherin commented on scripts/pypi_cleanup.py:
```diff
@@ -118,12 +119,18 @@ def run(self):
+
+            pkg_vers = []
+            for version_key, versions in version_dict.items():
+                if version_key in releases_by_date.keys():
```
Can we add some comments here about what this is exactly accomplishing?

Mytherin reviewed (commented):
Thanks for the fixes! Looks good - I tried it locally and it mostly works. Some comments:

Mytherin commented:
For testing, maybe we can expose the `--dry` parameter to this script. We don't need to login when not actually deleting, so that could then be run locally as well.

ZuleykhaPavlichenkova-TomTom committed:
remove retain_days and duckdb_version_base everywhere

ZuleykhaPavlichenkova-TomTom committed:
add a comment

ZuleykhaPavlichenkova-TomTom committed:
add --dry option for local testing
when it passed to the script, it will output the list of the PRE-RELEASE versions to delete and return

hmeriann commented:
I've added the `--dry` parameter for testing. Now it should output the list of names to delete and return, when `--dry` is passed to the script. Added the comment about that as well

ZuleykhaPavlichenkova-TomTom committed:
Merge remote-tracking branch 'upstream/main' into 4394-invokeci-python-twine-upload-had-failed-on-main

Mytherin commented:
Thanks! LGTM now. I've merged it into `main` since you rebased it off of main (perhaps accidentally?)

carlopi commented:
I think either is good

Nice the more solid rework, it should also make possible to just run this every day, without actually checking.

hmeriann commented:
> Thanks! LGTM now. I've merged it into `main` since you rebased it off of main (perhaps accidentally?) 

Thanks! Yes, it was accidentally 🙈 I was going to update a branch, but not this one
